grblas/_agg.py

```diff
@@ -517,35 +511,32 @@ def _argminmax(agg, updater, expr, *, in_composite, monoid):
 )
 
 
-def _first_last(agg, updater, expr, *, in_composite, semiring):
+def _first_last(agg, updater, expr, *, in_composite, semiring_):
     if expr.cfunc_name == "GrB_Matrix_reduce_Aggregator":
         A = expr.args[0]
         if expr.method_name == "reduce_columnwise":
             A = A.T
         init = expr._new_vector(bool, size=A._ncols)
         init[...] = False  # O(1) dense vector in SuiteSparse 5
-        step1 = semiring(A @ init).new()
+        step1 = semiring_(A @ init).new()
         Is, Js = step1.to_values()
-        # TODO: perform these loops in e.g. Cython
-        # Populate numpy array
-        vals = np.empty(Is.size, dtype=A.dtype.np_type)
-        for index, (i, j) in enumerate(zip(Is, Js)):
-            vals[index] = A[i, j].new().value
-        # or Vector
-        # v = expr._new_vector(A.dtype, size=A._nrows)
-        # for i, j in zip(Is, Js):
-        #     v[i] = A[i, j].new().value
-        result = Vector.from_values(Is, vals, size=A._nrows)
+
+        Matrix_ = type(expr._new_matrix(bool))
+        P = Matrix_.from_values(Js, Is, 1, nrows=A._ncols, ncols=A._nrows)
+        mask = Matrix_.from_values(Is, Is, True, nrows=A._nrows, ncols=A._nrows)
+        result = semiring.any_first(A @ P).new(mask=mask.S).diag()
+
         updater << result
         if in_composite:
             return updater.parent
     elif expr.cfunc_name.startswith("GrB_Vector_reduce"):
         v = expr.args[0]
         init = expr._new_matrix(bool, nrows=v._size, ncols=1)
         init[...] = False  # O(1) dense matrix in SuiteSparse 5
-        step1 = semiring(v @ init).new()
+        step1 = semiring_(v @ init).new()
         index = step1[0].new().value
-        if index is None:
+        # `==` instead of `is` automatically triggers index.compute() in dask-grblas:
+        if index == None:  # noqa
```

Comment:
We might want to consider supporting e.g. `t = v[s].new()`  where `s` and `t` are both `gb.Scalar`s that are empty.  Currently, scalar indices are `GrB_Index` type, not `GrB_Scalar` type, so this isn't supported by SuiteSparse:GraphBLAS.  But, we could support this in `grblas` and `dask-grblas`, and doing so would allow us to avoid `dask-grblas` computing during `index == None`.

Reply:
Great idea from an optimization standpoint!   

Heh, this automatically means also dealing with lists of indices of different types, like `v[2 s 1 None t].new()`.